java.lang.ExceptionInInitializerError at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101)

I´m having that exception for just creating a Cloner object:

Cloner cloner = new Cloner();

I'm usgin Android Studio 3.5 in a project with target sdk API 29

Thanks!

Sure. Unfortunately it seems related to the Android API version in use, because it fails only for certain versions (>= Android 9)

2019-09-03 08:38:57.213 3567-3567/com.lulisoft.dreamteam E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.lulisoft.dreamteam, PID: 3567
    java.lang.ExceptionInInitializerError
        at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101)
        at com.rits.cloning.Cloner.init(Cloner.java:88)
        at com.rits.cloning.Cloner.<init>(Cloner.java:58)
        at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418)
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195)
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078)
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117)
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408)
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366)
        at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243)
        at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654)
        at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146)
        at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244)
        at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669)
        at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631)
        at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086)
        at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097)
        at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291)
        at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102)
        at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432)
        at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6)
        at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4)
        at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6680)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: com.rits.cloning.CloningException: java.lang.NoSuchFieldException: No field m in class Ljava/util/TreeSet; (declaration of 'java.util.TreeSet' appears in /system/framework/core-oj.jar)
        at com.rits.cloning.FastClonerTreeSet.<clinit>(FastClonerTreeSet.java:22)
        at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101) 
        at com.rits.cloning.Cloner.init(Cloner.java:88) 
        at com.rits.cloning.Cloner.<init>(Cloner.java:58) 
        at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144) 
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418) 
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195) 
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078) 
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117) 
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408) 
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366) 
        at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243) 
        at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654) 
        at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146) 
        at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244) 
        at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669) 
        at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631) 
        at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086) 
        at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097) 
        at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291) 
        at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102) 
        at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432) 
        at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6) 
        at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4) 
        at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121) 
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6680) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     Caused by: java.lang.NoSuchFieldException: No field m in class Ljava/util/TreeSet; (declaration of 'java.util.TreeSet' appears in /system/framework/core-oj.jar)
        at java.lang.Class.getDeclaredField(Native Method)
        at com.rits.cloning.FastClonerTreeSet.<clinit>(FastClonerTreeSet.java:17)
        at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101) 
        at com.rits.cloning.Cloner.init(Cloner.java:88) 
        at com.rits.cloning.Cloner.<init>(Cloner.java:58) 
        at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144) 
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418) 
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195) 
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078) 
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117) 
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408) 
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366) 
        at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243) 
        at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654) 
        at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146) 
        at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244) 
        at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669) 
        at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631) 
        at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086) 
        at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097) 
        at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291) 
        at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102) 
        at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432) 
        at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6) 
        at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4) 
        at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121) 
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6680) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

> Sure. Unfortunately it seems related to the Android API version in use, because it fails only for certain versions (>= Android 9)
> 
> 2019-09-03 08:38:57.213 3567-3567/com.lulisoft.dreamteam E/AndroidRuntime: FATAL EXCEPTION: main
> Process: com.lulisoft.dreamteam, PID: 3567
> java.lang.ExceptionInInitializerError
> at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101)
> at com.rits.cloning.Cloner.init(Cloner.java:88)
> at com.rits.cloning.Cloner.(Cloner.java:58)
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144)
> at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414)
> at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418)
> at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195)
> at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078)
> at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117)
> at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408)
> at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366)
> at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243)
> at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654)
> at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146)
> at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244)
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669)
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631)
> at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086)
> at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097)
> at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291)
> at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102)
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432)
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6)
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4)
> at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121)
> at android.os.Handler.handleCallback(Handler.java:873)
> at android.os.Handler.dispatchMessage(Handler.java:99)
> at android.os.Looper.loop(Looper.java:193)
> at android.app.ActivityThread.main(ActivityThread.java:6680)
> at java.lang.reflect.Method.invoke(Native Method)
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
> Caused by: com.rits.cloning.CloningException: java.lang.NoSuchFieldException: No field m in class Ljava/util/TreeSet; (declaration of 'java.util.TreeSet' appears in /system/framework/core-oj.jar)
> at com.rits.cloning.FastClonerTreeSet.(FastClonerTreeSet.java:22)
> at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101) 
> at com.rits.cloning.Cloner.init(Cloner.java:88) 
> at com.rits.cloning.Cloner.(Cloner.java:58) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144) 
> at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
> at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418) 
> at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195) 
> at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078) 
> at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117) 
> at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408) 
> at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366) 
> at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243) 
> at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654) 
> at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146) 
> at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631) 
> at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086) 
> at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097) 
> at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291) 
> at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4) 
> at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121) 
> at android.os.Handler.handleCallback(Handler.java:873) 
> at android.os.Handler.dispatchMessage(Handler.java:99) 
> at android.os.Looper.loop(Looper.java:193) 
> at android.app.ActivityThread.main(ActivityThread.java:6680) 
> at java.lang.reflect.Method.invoke(Native Method) 
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
> Caused by: java.lang.NoSuchFieldException: No field m in class Ljava/util/TreeSet; (declaration of 'java.util.TreeSet' appears in /system/framework/core-oj.jar)
> at java.lang.Class.getDeclaredField(Native Method)
> at com.rits.cloning.FastClonerTreeSet.(FastClonerTreeSet.java:17)
> at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101) 
> at com.rits.cloning.Cloner.init(Cloner.java:88) 
> at com.rits.cloning.Cloner.(Cloner.java:58) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144) 
> at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
> at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418) 
> at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195) 
> at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078) 
> at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117) 
> at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408) 
> at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366) 
> at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243) 
> at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654) 
> at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146) 
> at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631) 
> at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086) 
> at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097) 
> at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291) 
> at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4) 
> at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121) 
> at android.os.Handler.handleCallback(Handler.java:873) 
> at android.os.Handler.dispatchMessage(Handler.java:99) 
> at android.os.Looper.loop(Looper.java:193) 
> at android.app.ActivityThread.main(ActivityThread.java:6680) 
> at java.lang.reflect.Method.invoke(Native Method) 
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

可能是因为反射的导致。
但是我认为这里不需要用到反射，所以更改这个类之后就可以了。

`public class FastClonerTreeSet implements IFastCloner {
//	private static final Field m;
//	private static final Field comparator;

//	static {
//		try {
//			m = TreeSet.class.getDeclaredField("m");
//			m.setAccessible(true);
//			comparator = TreeMap.class.getDeclaredField("comparator");
//			comparator.setAccessible(true);
//		} catch (NoSuchFieldException e) {
//			throw new CloningException(e);
//		}
//	}

	@SuppressWarnings("unchecked")
	public Object clone(Object t, IDeepCloner cloner, Map<Object, Object> clones) {
//		TreeSet treeSet = (TreeSet) t;
//		TreeSet result = null;
//		try {
//			result = new TreeSet((Comparator) comparator.get(m.get(t)));
//		} catch (IllegalAccessException e) {
//			throw new CloningException("Failed to get the comparator from a tree set", e);
//		}
//		for (Object o : treeSet) {
//			result.add(cloner.deepClone(o, clones));
//		}

		final TreeSet treeSet = (TreeSet) t;
		final TreeSet result = new TreeSet(treeSet.comparator());
		for(final Object o: treeSet){
			result.add(cloner.deepClone(o, clones));
		}

		return result;
	}
}`

> Sure. Unfortunately it seems related to the Android API version in use, because it fails only for certain versions (>= Android 9)
> 
> 2019-09-03 08:38:57.213 3567-3567/com.lulisoft.dreamteam E/AndroidRuntime: FATAL EXCEPTION: main
> Process: com.lulisoft.dreamteam, PID: 3567
> java.lang.ExceptionInInitializerError
> at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101)
> at com.rits.cloning.Cloner.init(Cloner.java:88)
> at com.rits.cloning.Cloner.(Cloner.java:58)
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144)
> at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414)
> at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418)
> at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195)
> at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078)
> at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117)
> at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408)
> at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366)
> at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243)
> at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654)
> at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146)
> at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244)
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669)
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631)
> at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086)
> at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097)
> at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291)
> at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102)
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432)
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6)
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4)
> at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121)
> at android.os.Handler.handleCallback(Handler.java:873)
> at android.os.Handler.dispatchMessage(Handler.java:99)
> at android.os.Looper.loop(Looper.java:193)
> at android.app.ActivityThread.main(ActivityThread.java:6680)
> at java.lang.reflect.Method.invoke(Native Method)
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
> Caused by: com.rits.cloning.CloningException: java.lang.NoSuchFieldException: No field m in class Ljava/util/TreeSet; (declaration of 'java.util.TreeSet' appears in /system/framework/core-oj.jar)
> at com.rits.cloning.FastClonerTreeSet.(FastClonerTreeSet.java:22)
> at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101) 
> at com.rits.cloning.Cloner.init(Cloner.java:88) 
> at com.rits.cloning.Cloner.(Cloner.java:58) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144) 
> at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
> at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418) 
> at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195) 
> at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078) 
> at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117) 
> at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408) 
> at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366) 
> at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243) 
> at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654) 
> at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146) 
> at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631) 
> at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086) 
> at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097) 
> at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291) 
> at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4) 
> at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121) 
> at android.os.Handler.handleCallback(Handler.java:873) 
> at android.os.Handler.dispatchMessage(Handler.java:99) 
> at android.os.Looper.loop(Looper.java:193) 
> at android.app.ActivityThread.main(ActivityThread.java:6680) 
> at java.lang.reflect.Method.invoke(Native Method) 
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
> Caused by: java.lang.NoSuchFieldException: No field m in class Ljava/util/TreeSet; (declaration of 'java.util.TreeSet' appears in /system/framework/core-oj.jar)
> at java.lang.Class.getDeclaredField(Native Method)
> at com.rits.cloning.FastClonerTreeSet.(FastClonerTreeSet.java:17)
> at com.rits.cloning.Cloner.registerFastCloners(Cloner.java:101) 
> at com.rits.cloning.Cloner.init(Cloner.java:88) 
> at com.rits.cloning.Cloner.(Cloner.java:58) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailCuadrosFragment.onCreate(PartidoDetailCuadrosFragment.java:144) 
> at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
> at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1418) 
> at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1195) 
> at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1078) 
> at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:117) 
> at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2408) 
> at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366) 
> at androidx.fragment.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2243) 
> at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:654) 
> at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:146) 
> at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:669) 
> at androidx.viewpager.widget.ViewPager.setCurrentItemInternal(ViewPager.java:631) 
> at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1086) 
> at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097) 
> at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291) 
> at com.lulisoft.dreamteam.tabPartidos.CambiosPagerAdapter.add(CambiosPagerAdapter.java:102) 
> at com.lulisoft.dreamteam.tabPartidos.PartidoDetailDatosFragment.lambda$null$15$PartidoDetailDatosFragment(PartidoDetailDatosFragment.java:432) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:6) 
> at com.lulisoft.dreamteam.tabPartidos.-$$Lambda$PartidoDetailDatosFragment$-Xs0Ha6cYdNtNtrmLVb8gcuCjGQ.done(Unknown Source:4) 
> at com.parse.ParseTaskUtils$2$1.run(ParseTaskUtils.java:121) 
> at android.os.Handler.handleCallback(Handler.java:873) 
> at android.os.Handler.dispatchMessage(Handler.java:99) 
> at android.os.Looper.loop(Looper.java:193) 
> at android.app.ActivityThread.main(ActivityThread.java:6680) 
> at java.lang.reflect.Method.invoke(Native Method) 
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

是因为反射导致的。
我认为这里不需要反射，所以去掉了就可以了
```
public class FastClonerTreeSet implements IFastCloner {
// private static final Field m;
// private static final Field comparator;

// static {
//    try {
//       m = TreeSet.class.getDeclaredField("m");
//       m.setAccessible(true);
//       comparator = TreeMap.class.getDeclaredField("comparator");
//       comparator.setAccessible(true);
//    } catch (NoSuchFieldException e) {
//       throw new CloningException(e);
//    }
// }

   @SuppressWarnings("unchecked")
   public Object clone(Object t, IDeepCloner cloner, Map<Object, Object> clones) {
//    TreeSet treeSet = (TreeSet) t;
//    TreeSet result = null;
//    try {
//       result = new TreeSet((Comparator) comparator.get(m.get(t)));
//    } catch (IllegalAccessException e) {
//       throw new CloningException("Failed to get the comparator from a tree set", e);
//    }
//    for (Object o : treeSet) {
//       result.add(cloner.deepClone(o, clones));
//    }

      final TreeSet treeSet = (TreeSet) t;
      final TreeSet result = new TreeSet(treeSet.comparator());
      for(final Object o: treeSet){
         result.add(cloner.deepClone(o, clones));
      }

      return result;
   }
}
```

--- a/src/main/java/com/rits/cloning/FastClonerTreeSet.java
+++ b/src/main/java/com/rits/cloning/FastClonerTreeSet.java
@@ -2,36 +2,15 @@ package com.rits.cloning;
 
 import com.rits.cloning.IFastCloner;
 
-import java.lang.reflect.Field;
-import java.util.Comparator;
 import java.util.Map;
-import java.util.TreeMap;
 import java.util.TreeSet;
 
 public class FastClonerTreeSet implements IFastCloner {
-	private static final Field m;
-	private static final Field comparator;
-
-	static {
-		try {
-			m = TreeSet.class.getDeclaredField("m");
-			m.setAccessible(true);
-			comparator = TreeMap.class.getDeclaredField("comparator");
-			comparator.setAccessible(true);
-		} catch (NoSuchFieldException e) {
-			throw new CloningException(e);
-		}
-	}
-
+	@Override
 	@SuppressWarnings("unchecked")
 	public Object clone(Object t, IDeepCloner cloner, Map<Object, Object> clones) {
-		TreeSet treeSet = (TreeSet) t;
-		TreeSet result = null;
-		try {
-			result = new TreeSet((Comparator) comparator.get(m.get(t)));
-		} catch (IllegalAccessException e) {
-			throw new CloningException("Failed to get the comparator from a tree set", e);
-		}
+		TreeSet<?> treeSet = (TreeSet<?>) t;
+		TreeSet result = new TreeSet(treeSet.comparator());
 		for (Object o : treeSet) {
 			result.add(cloner.deepClone(o, clones));
 		}

